File: packages/python/src/vendorval_sdk/resources/_certifications.py
"""Certifications resource (sync + async). Phase N customer-facing
reshape, Workstream B.

Today this surface is read-only — `list` + `retrieve`. POST + DELETE
(manual upload + revoke) land in a follow-up SDK release once those
API routes ship.
"""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

from .._models import Response
from .._request import ResolvedConfig, execute_async, execute_sync, prepare


def _build_query(
    *,
    entity_id: str | None,
    tin: str | None,
    uei: str | None,
    duns: str | None,
    lei: str | None,
    vat_id: str | None,
    state_entity_id: str | None,
    npi: str | None,
    issuer: str | None,
    status: str | None,
    scope: str | list[str] | None,
    expiring_within_days: int | None,
    limit: int | None,
    offset: int | None,
) -> dict[str, Any]:
    """Build the GET query payload.

    Identifier params (tin / uei / duns / lei / vat_id / state_entity_id /
    npi) are normalized + hashed + joined server-side via the same path
    `/v1/entities/lookup` uses. Saves callers a 2-step lookup-then-query
    flow. Tenant-scoped at the API; passing multiple identifiers that
    resolve to different entities → 400.

    `scope` is Phase 5 of data #155 — comma-separated multi-select on
    the awarding authority's coarse scope. Pass a single value
    (e.g. `'federal'`) or a list — the SDK joins lists with `,` for
    the api's wire format.
    """
    return {
        "entity_id": entity_id,
        "tin": tin,
        "uei": uei,
        "duns": duns,
        "lei": lei,
        "vat_id": vat_id,
        "state_entity_id": state_entity_id,
        "npi": npi,
        "issuer": issuer,
        "status": status,
        "scope": ",".join(scope) if isinstance(scope, list) else scope,
        "expiring_within_days": expiring_within_days,
        "limit": limit,
        "offset": offset,
    }


class CertificationsResource:
    def __init__(self, cfg: ResolvedConfig, client: httpx.Client) -> None:
        self._cfg = cfg
        self._client = client

    def list(
        self,
        *,
        entity_id: str | None = None,
        tin: str | None = None,
        uei: str | None = None,
        duns: str | None = None,
        lei: str | None = None,
        vat_id: str | None = None,
        state_entity_id: str | None = None,
        npi: str | None = None,
        issuer: str | None = None,
        status: str | None = None,
        scope: str | list[str] | None = None,
        expiring_within_days: int | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Response:
        """List certifications for the calling org.

        Returns the full list envelope verbatim so callers see pagination
        metadata (`total`, `has_more`, `limit`, `offset`) without
        re-querying for the count. Access rows via `response["data"]`,
        or call `response.to_dict()` to work with the full payload as a
        plain dictionary.

        Identifier params (`tin`, `uei`, `duns`, `lei`, `vat_id`,
        `state_entity_id`, `npi`) scope to the entity that matches the
        identifier — server normalizes + hashes + joins the same way
        `/v1/entities/lookup` does. Saves a 2-step lookup-then-query flow.
        """
        prepared = prepare(
            self._cfg,
            method="GET",
            path="/v1/certifications",
            query=_build_query(
                entity_id=entity_id,
                tin=tin,
                uei=uei,
                duns=duns,
                lei=lei,
                vat_id=vat_id,
                state_entity_id=state_entity_id,
                npi=npi,
                issuer=issuer,
                status=status,
                scope=scope,
                expiring_within_days=expiring_within_days,
                limit=limit,
                offset=offset,
            ),
        )
        res = execute_sync(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    def retrieve(self, certification_id: str) -> Response:
        """Fetch a single certification by its public id (`cert_…`)."""
        prepared = prepare(
            self._cfg,
            method="GET",
            path=f"/v1/certifications/{quote(certification_id, safe='')}",
        )
        res = execute_sync(self._client, prepared)
        return Response(res.data, res.request_id, res.status)


class AsyncCertificationsResource:
    def __init__(self, cfg: ResolvedConfig, client: httpx.AsyncClient) -> None:
        self._cfg = cfg
        self._client = client

    async def list(
        self,
        *,
        entity_id: str | None = None,
        tin: str | None = None,
        uei: str | None = None,
        duns: str | None = None,
        lei: str | None = None,
        vat_id: str | None = None,
        state_entity_id: str | None = None,
        npi: str | None = None,
        issuer: str | None = None,
        status: str | None = None,
        scope: str | list[str] | None = None,
        expiring_within_days: int | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Response:
        prepared = prepare(
            self._cfg,
            method="GET",
            path="/v1/certifications",
            query=_build_query(
                entity_id=entity_id,
                tin=tin,
                uei=uei,
                duns=duns,
                lei=lei,
                vat_id=vat_id,
                state_entity_id=state_entity_id,
                npi=npi,
                issuer=issuer,
                status=status,
                scope=scope,
                expiring_within_days=expiring_within_days,
                limit=limit,
                offset=offset,
            ),
        )
        res = await execute_async(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    async def retrieve(self, certification_id: str) -> Response:
        prepared = prepare(
            self._cfg,
            method="GET",
            path=f"/v1/certifications/{quote(certification_id, safe='')}",
        )
        res = await execute_async(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

File: packages/node/src/resources/certifications.ts
import { performRequest, type ResolvedClientOptions } from "../request.js";
import type {
  Certification,
  CertificationsListParams,
  CertificationsListResponse,
} from "../types/shared.js";

/**
 * `client.certifications.*` — read access to entity credentials
 * (state MWBE, NMSDC, WBENC, ISO, SOC 2, etc.). Phase N customer-facing
 * reshape, Workstream B.
 *
 * Today this surface is read-only; POST + DELETE (manual upload + revoke)
 * land in a follow-up SDK release once those API routes ship.
 */
export class CertificationsResource {
  constructor(private readonly client: ResolvedClientOptions) {}

  /**
   * List certifications for the calling org. Filters narrow the result
   * set on the server. Returns the full list envelope verbatim so
   * callers see pagination metadata (`total`, `has_more`, `limit`,
   * `offset`) without re-querying for the count.
   *
   * Async iteration helper:
   *
   *   for (const cert of (await client.certifications.list()).data) { … }
   *
   * Cursor pagination is a planned follow-up — when it lands, this
   * shape will accept a `cursor` param + return `next_cursor` without
   * a breaking change.
   */
  async list(
    params: CertificationsListParams = {},
  ): Promise<CertificationsListResponse & { _requestId: string | null }> {
    const query: Record<string, string | number | boolean | undefined> = {};
    if (params.entity_id !== undefined) query.entity_id = params.entity_id;
    // Identifier-resolved scoping — server normalizes + hashes + joins.
    if (params.tin !== undefined) query.tin = params.tin;
    if (params.uei !== undefined) query.uei = params.uei;
    if (params.duns !== undefined) query.duns = params.duns;
    if (params.lei !== undefined) query.lei = params.lei;
    if (params.vat_id !== undefined) query.vat_id = params.vat_id;
    if (params.state_entity_id !== undefined) query.state_entity_id = params.state_entity_id;
    if (params.npi !== undefined) query.npi = params.npi;
    if (params.issuer !== undefined) query.issuer = params.issuer;
    if (params.status !== undefined) query.status = params.status;
    // Phase 5 of data #155 — `?scope=` is comma-separated multi-select
    // at the wire level. SDK consumers can pass a single value or an
    // array; we join the array here so the typing stays ergonomic.
    if (params.scope !== undefined) {
      query.scope = Array.isArray(params.scope) ? params.scope.join(",") : params.scope;
    }
    if (params.expiring_within_days !== undefined) {
      query.expiring_within_days = params.expiring_within_days;
    }
    if (params.limit !== undefined) query.limit = params.limit;
    if (params.offset !== undefined) query.offset = params.offset;

    const res = await performRequest<CertificationsListResponse>(this.client, {
      method: "GET",
      path: "/v1/certifications",
      query,
    });

    return { ...res.data, _requestId: res.requestId };
  }

  /**
   * Fetch a single certification by its public id (`cert_…`).
   * Throws a NotFound error if the id doesn't resolve under the
   * caller's org.
   */
  async retrieve(id: string): Promise<Certification & { _requestId: string | null }> {
    const res = await performRequest<Certification>(this.client, {
      method: "GET",
      path: `/v1/certifications/${encodeURIComponent(id)}`,
    });
    return { ...res.data, _requestId: res.requestId };
  }
}
